plugins/org.jkiss.dbeaver.model.ai/src/org/jkiss/dbeaver/model/ai/client/GPTAPIClient.java

```diff
@@ -0,0 +1,148 @@
+/*
+ * DBeaver - Universal Database Manager
+ * Copyright (C) 2010-2023 DBeaver Corp and others
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jkiss.dbeaver.model.ai.client;
+
+import com.theokanning.openai.OpenAiService;
+import com.theokanning.openai.completion.CompletionChoice;
+import com.theokanning.openai.completion.CompletionRequest;
+import org.eclipse.core.runtime.IProgressMonitor;
+import org.jkiss.code.NotNull;
+import org.jkiss.dbeaver.DBException;
+import org.jkiss.dbeaver.model.ai.GPTPreferences;
+import org.jkiss.dbeaver.model.ai.internal.GTPActivator;
+import org.jkiss.dbeaver.model.ai.formatter.GPTRequestFormatter;
+import org.jkiss.dbeaver.model.preferences.DBPPreferenceStore;
+import org.jkiss.dbeaver.model.runtime.DBRProgressMonitor;
+import org.jkiss.dbeaver.model.secret.DBSSecretController;
+import org.jkiss.dbeaver.model.struct.DBSObject;
+import org.jkiss.dbeaver.runtime.DBWorkbench;
+import org.jkiss.utils.CommonUtils;
+import retrofit2.HttpException;
+
+
+import java.time.Duration;
+import java.util.List;
+import java.util.Optional;
+
+public class GPTAPIClient {
+    //How many retries may be done if code 429 happens
+    private static final int MAX_REQUEST_ATTEMPTS = 3;
+
+    private static OpenAiService CLIENT_INSTANCE = null;
```

Comment:
It should not be a singleton.

1. Extract methods into an abstract interface and create an internal implementation
2. Provide a service (`org.jkiss.dbeaver.service`) that will provide an implementation for that interface